server/src/main/java/io/deephaven/server/arrow/ArrowFlightUtil.java

```diff
@@ -66,6 +71,166 @@ public class ArrowFlightUtil {
+                                    msg.rowsIncluded, columns));
+                        }
+
+                        if (msg.rowsIncluded.size() > 0) {
+                            // very simplistic logic to take the last snapshot and extrapolate max
+                            // number of rows that will not exceed the target UGP processing time
+                            // percentage
+                            long targetNanos = (long) (TARGET_SNAPSHOT_PERCENTAGE
+                                    * UpdateGraphProcessor.DEFAULT.getTargetCycleDurationMillis()
+                                    * 1000000);
+
+                            long nanosPerCell = elapsed / (msg.rowsIncluded.size() * columnCount);
+
+                            // apply an exponential moving average to filter the data
+                            if (snapshotNanosPerCell == 0) {
+                                snapshotNanosPerCell = nanosPerCell; // initialize to first value
+                            } else {
+                                // EMA smoothing factor is 0.1 (N = 10)
+                                snapshotNanosPerCell =
+                                        (snapshotNanosPerCell * 0.9) + (nanosPerCell * 0.1);
+                            }
+
+                            snapshotTargetCellCount =
+                                    (long) (targetNanos / Math.max(1, snapshotNanosPerCell));
+                        }
+                    }
+                    // increment the starting key
+                    startKey = endKey + 1;
+                }
+            }
+        } else {
+            // get ourselves some data (reversing viewport as instructed)
+            final BarrageMessage msg;
+
+            final long snapshotStartTm = System.nanoTime();
+            if (reverseViewport) {
+                msg = ConstructSnapshot.constructBackplaneSnapshotInPositionSpace(log, table,
+                        columns, null, viewport);
+            } else {
+                msg = ConstructSnapshot.constructBackplaneSnapshotInPositionSpace(log, table,
+                        columns, viewport, null);
+            }
+            metrics.snapshotNanos = System.nanoTime() - snapshotStartTm;
+
+            msg.modColumnData = ZERO_MOD_COLUMNS; // no mod column data
+
+            // translate the viewport to keyspace and make the call
+            try (final BarrageStreamGenerator bsg = new BarrageStreamGenerator(msg, metrics);
+                    final RowSet keySpaceViewport =
+                            viewport != null
+                                    ? msg.rowsAdded.subSetForPositions(viewport, reverseViewport)
+                                    : null) {
+                listener.onNext(
+                        bsg.getSnapshotView(snapshotRequestOptions, viewport,
+                                reverseViewport, keySpaceViewport, columns));
+            }
+        }
+    }
```

Comment:
A minor nit on the method structure. The `if`/`else` feels like the majority of the method and therefore a less readable indentation. Let's move the iterative implementation into a helper, return, and drop the `else` block. Something like this:

```java
if (!table.isRefreshing()) {
    createAndSendStaticSnapshot(...);
    return;
}

// the original snapshot implementation that fetches a single BarrageMessage for the entire request
```